File: app/src/main/java/com/screenshot/ocr/ChatDetector.kt
package com.screenshot.ocr

import com.screenshot.ocr.models.ChatMessage
import com.screenshot.ocr.models.MessageAlignment

/**
 * Detects and formats chat conversations from extracted text
 */
class ChatDetector {

    private val chatPatterns = listOf(
        // "[ALIGNMENT] Name: message" - with alignment marker
        Regex("""^\[(LEFT|RIGHT)\]\s*([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$"""),
        // "[ALIGNMENT] [Time] Name: message"
        Regex("""^\[(LEFT|RIGHT)\]\s*([A-Z][a-zA-Z\s]{0,30})\s*[\[\(]([0-9:APM\s]+)[\]\)]\s*:\s*(.+)$"""),
        // "[ALIGNMENT] Name [Time]: message"
        Regex("""^\[(LEFT|RIGHT)\]\s*([A-Z][a-zA-Z\s]{0,30})\s*[\[\(]([0-9:APM\s]+)[\]\)]\s*:\s*(.+)$"""),
        // "Name: message" - without alignment marker (fallback)
        Regex("""^([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$"""),
        // "[Time] Name: message" or "(Time) Name: message"
        Regex("""^[\[\(]([0-9:APM\s]+)[\]\)]\s*([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$"""),
        // "Name [Time]: message"
        Regex("""^([A-Z][a-zA-Z\s]{0,30})\s*[\[\(]([0-9:APM\s]+)[\]\)]\s*:\s*(.+)$"""),
        // "HH:MM Name: message"
        Regex("""^([0-9]{1,2}:[0-9]{2}\s*(?:AM|PM)?)\s+([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$"""),
        // WhatsApp/Telegram: "[HH:MM] Name:" followed by message on next line
        Regex("""^[\[\(]([0-9]{1,2}:[0-9]{2})[\]\)]\s*([A-Z][a-zA-Z\s]+):?\s*$""")
    )

    private val quotePattern = Regex("""\[QUOTE:\s*([^\]]+)\]\s*(.*?)\s*\[/QUOTE\]""", RegexOption.DOT_MATCHES_ALL)

    private val timePatterns = listOf(
        Regex("""\b([0-9]{1,2}:[0-9]{2}\s*(?:AM|PM)?)\b"""),
        Regex("""\b([0-9]{1,2}:[0-9]{2}:[0-9]{2})\b""")
    )

    /**
     * Determine if text is likely a chat conversation
     */
    fun isLikelyChat(text: String): Boolean {
        val lines = text.lines().filter { it.isNotBlank() }
        if (lines.size < 3) return false

        var chatIndicators = 0
        val totalLines = lines.size

        for (line in lines) {
            // Check for chat patterns
            if (chatPatterns.any { it.matches(line) }) {
                chatIndicators++
            }

            // Check for timestamps
            if (timePatterns.any { it.containsMatchIn(line) }) {
                chatIndicators++
            }
        }

        // If more than 30% of lines match chat patterns
        return (chatIndicators.toFloat() / totalLines) > 0.3f
    }

    /**
     * Extract chat messages from text
     */
    fun extractMessages(text: String): List<ChatMessage> {
        val lines = text.lines()
        val messages = mutableListOf<ChatMessage>()
        var currentMessage: ChatMessage? = null

        for ((lineNum, line) in lines.withIndex()) {
            val trimmedLine = line.trim()
            if (trimmedLine.isEmpty()) continue

            var matched = false

            // Pattern 0: "[ALIGNMENT] Name: message"
            chatPatterns[0].matchEntire(trimmedLine)?.let { match ->
                val (alignmentStr, speaker, messageContent) = match.destructured
                val alignment = when (alignmentStr.uppercase()) {
                    "LEFT" -> MessageAlignment.LEFT
                    "RIGHT" -> MessageAlignment.RIGHT
                    else -> MessageAlignment.UNKNOWN
                }

                val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                if (speaker.length in 1..30) {
                    currentMessage = ChatMessage(
                        speaker = speaker.trim(),
                        message = actualMessage.trim(),
                        lineNumber = lineNum,
                        alignment = alignment,
                        quotedSpeaker = quotedSpeaker,
                        quotedMessage = quotedMsg
                    )
                    messages.add(currentMessage!!)
                    matched = true
                }
            }

            // Pattern 1: "[ALIGNMENT] Name [Time]: message"
            if (!matched) {
                chatPatterns[1].matchEntire(trimmedLine)?.let { match ->
                    val groups = match.groupValues
                    val alignmentStr = groups.getOrNull(1) ?: ""
                    val speaker = groups.getOrNull(2) ?: ""
                    val timestamp = groups.getOrNull(3) ?: ""
                    val messageContent = groups.getOrNull(4) ?: ""

                    val alignment = when (alignmentStr.uppercase()) {
                        "LEFT" -> MessageAlignment.LEFT
                        "RIGHT" -> MessageAlignment.RIGHT
                        else -> MessageAlignment.UNKNOWN
                    }

                    val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                    currentMessage = ChatMessage(
                        speaker = speaker.trim(),
                        message = actualMessage.trim(),
                        timestamp = timestamp.trim(),
                        lineNumber = lineNum,
                        alignment = alignment,
                        quotedSpeaker = quotedSpeaker,
                        quotedMessage = quotedMsg
                    )
                    messages.add(currentMessage!!)
                    matched = true
                }
            }

            // Pattern 2: "[ALIGNMENT] Name [Time]: message" (alternative format)
            if (!matched) {
                chatPatterns[2].matchEntire(trimmedLine)?.let { match ->
                    val groups = match.groupValues
                    val alignmentStr = groups.getOrNull(1) ?: ""
                    val speaker = groups.getOrNull(2) ?: ""
                    val timestamp = groups.getOrNull(3) ?: ""
                    val messageContent = groups.getOrNull(4) ?: ""

                    val alignment = when (alignmentStr.uppercase()) {
                        "LEFT" -> MessageAlignment.LEFT
                        "RIGHT" -> MessageAlignment.RIGHT
                        else -> MessageAlignment.UNKNOWN
                    }

                    val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                    currentMessage = ChatMessage(
                        speaker = speaker.trim(),
                        message = actualMessage.trim(),
                        timestamp = timestamp.trim(),
                        lineNumber = lineNum,
                        alignment = alignment,
                        quotedSpeaker = quotedSpeaker,
                        quotedMessage = quotedMsg
                    )
                    messages.add(currentMessage!!)
                    matched = true
                }
            }

            // Pattern 3: "Name: message" (fallback without alignment)
            if (!matched) {
                chatPatterns[3].matchEntire(trimmedLine)?.let { match ->
                    val (speaker, messageContent) = match.destructured
                    val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                    if (speaker.length in 1..30) {
                        currentMessage = ChatMessage(
                            speaker = speaker.trim(),
                            message = actualMessage.trim(),
                            lineNumber = lineNum,
                            quotedSpeaker = quotedSpeaker,
                            quotedMessage = quotedMsg
                        )
                        messages.add(currentMessage!!)
                        matched = true
                    }
                }
            }

            // Pattern 4: "[Time] Name: message"
            if (!matched) {
                chatPatterns[4].matchEntire(trimmedLine)?.let { match ->
                    val (timestamp, speaker, messageContent) = match.destructured
                    val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                    currentMessage = ChatMessage(
                        speaker = speaker.trim(),
                        message = actualMessage.trim(),
                        timestamp = timestamp.trim(),
                        lineNumber = lineNum,
                        quotedSpeaker = quotedSpeaker,
                        quotedMessage = quotedMsg
                    )
                    messages.add(currentMessage!!)
                    matched = true
                }
            }

            // Pattern 5: "Name [Time]: message"
            if (!matched) {
                chatPatterns[5].matchEntire(trimmedLine)?.let { match ->
                    val (speaker, timestamp, messageContent) = match.destructured
                    val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                    currentMessage = ChatMessage(
                        speaker = speaker.trim(),
                        message = actualMessage.trim(),
                        timestamp = timestamp.trim(),
                        lineNumber = lineNum,
                        quotedSpeaker = quotedSpeaker,
                        quotedMessage = quotedMsg
                    )
                    messages.add(currentMessage!!)
                    matched = true
                }
            }

            // Pattern 6: "HH:MM Name: message"
            if (!matched) {
                chatPatterns[6].matchEntire(trimmedLine)?.let { match ->
                    val (timestamp, speaker, messageContent) = match.destructured
                    val (quotedSpeaker, quotedMsg, actualMessage) = extractQuote(messageContent)

                    currentMessage = ChatMessage(
                        speaker = speaker.trim(),
                        message = actualMessage.trim(),
                        timestamp = timestamp.trim(),
                        lineNumber = lineNum,
                        quotedSpeaker = quotedSpeaker,
                        quotedMessage = quotedMsg
                    )
                    messages.add(currentMessage!!)
                    matched = true
                }
            }

            // If no pattern matched and we have a current message, it's a continuation
            if (!matched && currentMessage != null) {
                // Check if this looks like a continuation (doesn't start with alignment marker or capital + colon)
                if (!Regex("""^(\[(?:LEFT|RIGHT)\]\s*)?[A-Z][a-zA-Z\s]*:""").matches(trimmedLine)) {
                    currentMessage = currentMessage!!.copy(
                        message = currentMessage!!.message + "\n" + trimmedLine
                    )
                    // Update the message in the list
                    if (messages.isNotEmpty()) {
                        messages[messages.lastIndex] = currentMessage!!
                    }
                }
            }
        }

        return messages
    }

    /**
     * Extract quote information from a message
     * Returns (quotedSpeaker, quotedMessage, remainingMessage)
     */
    private fun extractQuote(messageContent: String): Triple<String?, String?, String> {
        val match = quotePattern.find(messageContent)
        return if (match != null) {
            val quotedSpeaker = match.groupValues.getOrNull(1)?.trim()
            val quotedMessage = match.groupValues.getOrNull(2)?.trim()
            val remainingMessage = messageContent.replace(match.value, "").trim()
            Triple(quotedSpeaker, quotedMessage, remainingMessage)
        } else {
            Triple(null, null, messageContent)
        }
    }

    /**
     * Format messages into a readable conversation
     */
    fun formatConversation(messages: List<ChatMessage>): String {
        if (messages.isEmpty()) return ""

        val sb = StringBuilder()
        sb.appendLine("=" * 60)
        sb.appendLine("CHAT CONVERSATION")
        sb.appendLine("=" * 60)
        sb.appendLine()

        // Get conversation statistics
        val speakers = messages.groupingBy { it.speaker }.eachCount()
        val alignmentCounts = messages.groupingBy { it.alignment }.eachCount()

        sb.appendLine("Participants: ${speakers.keys.joinToString(", ")}")
        sb.appendLine("Total messages: ${messages.size}")

        // Show alignment distribution if detected
        if (alignmentCounts.containsKey(MessageAlignment.LEFT) || alignmentCounts.containsKey(MessageAlignment.RIGHT)) {
            val leftSpeakers = messages.filter { it.alignment == MessageAlignment.LEFT }
                .map { it.speaker }.distinct()
            val rightSpeakers = messages.filter { it.alignment == MessageAlignment.RIGHT }
                .map { it.speaker }.distinct()

            if (leftSpeakers.isNotEmpty()) {
                sb.appendLine("Left-aligned: ${leftSpeakers.joinToString(", ")}")
            }
            if (rightSpeakers.isNotEmpty()) {
                sb.appendLine("Right-aligned: ${rightSpeakers.joinToString(", ")}")
            }
        }

        val quotedMessages = messages.count { it.quotedSpeaker != null }
        if (quotedMessages > 0) {
            sb.appendLine("Messages with quotes: $quotedMessages")
        }

        sb.appendLine()
        sb.appendLine("=" * 60)
        sb.appendLine()

        for (msg in messages) {
            // Format alignment indicator
            val alignmentIndicator = when (msg.alignment) {
                MessageAlignment.LEFT -> "◄ "
                MessageAlignment.RIGHT -> "► "
                MessageAlignment.UNKNOWN -> ""
            }

            // Format: [Alignment] [Time] Speaker:
            val header = buildString {
                append(alignmentIndicator)
                if (msg.timestamp != null) {
                    append("[${msg.timestamp}] ")
                }
                append("${msg.speaker}:")
            }

            sb.appendLine(header)

            // If there's a quote, show it first
            if (msg.quotedSpeaker != null && msg.quotedMessage != null) {
                sb.appendLine("  ┌─ Quoting ${msg.quotedSpeaker}:")
                msg.quotedMessage.lines().forEach { line ->
                    sb.appendLine("  │ $line")
                }
                sb.appendLine("  └─")
            }

            // Show the actual message
            msg.message.lines().forEach { line ->
                sb.appendLine("  $line")
            }

            sb.appendLine() // Empty line between messages
        }

        sb.appendLine("=" * 60)

        return sb.toString()
    }

    /**
     * Detect and format chat from text
     * Returns (isChat, messages)
     */
    fun detectAndFormat(text: String): Pair<Boolean, List<ChatMessage>> {
        val isChat = isLikelyChat(text)
        return if (isChat) {
            val messages = extractMessages(text)
            Pair(true, messages)
        } else {
            Pair(false, emptyList())
        }
    }

    private operator fun String.times(n: Int): String = this.repeat(n)
}

File: chat_detector.py
"""
Chat conversation detection and formatting module.
"""

import re
from typing import List, Dict, Tuple, Optional
from dataclasses import dataclass
from datetime import datetime


@dataclass
class ChatMessage:
    """Represents a single chat message."""
    speaker: str
    message: str
    timestamp: Optional[str] = None
    line_number: int = 0


class ChatDetector:
    """Detects and formats chat conversations from OCR text."""

    def __init__(self):
        """Initialize the chat detector with common patterns."""
        # Common chat patterns
        self.patterns = [
            # Pattern: "Name: message" or "Name : message"
            re.compile(r'^([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$'),
            # Pattern: "[Time] Name: message" or "(Time) Name: message"
            re.compile(r'^[\[\(]([0-9:APM\s]+)[\]\)]\s*([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$'),
            # Pattern: "Name [Time]: message"
            re.compile(r'^([A-Z][a-zA-Z\s]{0,30})\s*[\[\(]([0-9:APM\s]+)[\]\)]\s*:\s*(.+)$'),
            # Pattern: "Name (Time) message" (no colon)
            re.compile(r'^([A-Z][a-zA-Z\s]{0,30})\s*[\[\(]([0-9:APM\s]+)[\]\)]\s+(.+)$'),
            # Pattern: Messenger/WhatsApp style "Name" on one line, message on next
            re.compile(r'^([A-Z][a-zA-Z\s]{0,30})$'),
            # Pattern: Time-stamped messages "HH:MM Name: message"
            re.compile(r'^([0-9]{1,2}:[0-9]{2}\s*(?:AM|PM)?)\s+([A-Z][a-zA-Z\s]{0,30})\s*:\s*(.+)$'),
        ]

        # Common time patterns
        self.time_patterns = [
            re.compile(r'\b([0-9]{1,2}:[0-9]{2}\s*(?:AM|PM)?)\b'),
            re.compile(r'\b([0-9]{1,2}:[0-9]{2}:[0-9]{2})\b'),
        ]

    def is_likely_chat(self, text: str) -> bool:
        """
        Determine if the text is likely a chat conversation.

        Args:
            text: Input text

        Returns:
            True if text appears to be a chat conversation
        """
        lines = [line.strip() for line in text.split('\n') if line.strip()]

        if len(lines) < 3:
            return False

        chat_indicators = 0
        total_lines = len(lines)

        for line in lines:
            # Check for chat patterns
            for pattern in self.patterns[:3]:  # Check first 3 most common patterns
                if pattern.match(line):
                    chat_indicators += 1
                    break

            # Check for timestamps
            for time_pattern in self.time_patterns:
                if time_pattern.search(line):
                    chat_indicators += 0.5
                    break

        # If more than 30% of lines match chat patterns, it's likely a chat
        confidence = chat_indicators / total_lines
        return confidence > 0.3

    def extract_messages(self, text: str) -> List[ChatMessage]:
        """
        Extract chat messages from text.

        Args:
            text: Input text

        Returns:
            List of ChatMessage objects
        """
        lines = text.split('\n')
        messages = []
        current_message = None

        for i, line in enumerate(lines):
            line = line.strip()
            if not line:
                continue

            matched = False

            # Try each pattern
            # Pattern 1: "Name: message"
            match = self.patterns[0].match(line)
            if match and not matched:
                speaker, message = match.groups()
                if len(speaker) > 0 and len(speaker) < 30:
                    current_message = ChatMessage(
                        speaker=speaker.strip(),
                        message=message.strip(),
                        line_number=i
                    )
                    messages.append(current_message)
                    matched = True

            # Pattern 2: "[Time] Name: message"
            if not matched:
                match = self.patterns[1].match(line)
                if match:
                    timestamp, speaker, message = match.groups()
                    current_message = ChatMessage(
                        speaker=speaker.strip(),
                        message=message.strip(),
                        timestamp=timestamp.strip(),
                        line_number=i
                    )
                    messages.append(current_message)
                    matched = True

            # Pattern 3: "Name [Time]: message"
            if not matched:
                match = self.patterns[2].match(line)
                if match:
                    speaker, timestamp, message = match.groups()
                    current_message = ChatMessage(
                        speaker=speaker.strip(),
                        message=message.strip(),
                        timestamp=timestamp.strip(),
                        line_number=i
                    )
                    messages.append(current_message)
                    matched = True

            # Pattern 5: "HH:MM Name: message"
            if not matched:
                match = self.patterns[5].match(line)
                if match:
                    timestamp, speaker, message = match.groups()
                    current_message = ChatMessage(
                        speaker=speaker.strip(),
                        message=message.strip(),
                        timestamp=timestamp.strip(),
                        line_number=i
                    )
                    messages.append(current_message)
                    matched = True

            # If no pattern matched and we have a current message, it might be continuation
            if not matched and current_message:
                # Check if this looks like a continuation (doesn't start with capital + colon)
                if not re.match(r'^[A-Z][a-zA-Z\s]*:', line):
                    current_message.message += '\n' + line

        return messages

    def format_conversation(self, messages: List[ChatMessage],
                           include_timestamps: bool = True) -> str:
        """
        Format chat messages into a readable conversation.

        Args:
            messages: List of ChatMessage objects
            include_timestamps: Whether to include timestamps

        Returns:
            Formatted conversation string
        """
        if not messages:
            return ""

        formatted = []
        formatted.append("=" * 60)
        formatted.append("CHAT CONVERSATION")
        formatted.append("=" * 60)
        formatted.append("")

        for msg in messages:
            # Format: [Time] Speaker: Message
            if include_timestamps and msg.timestamp:
                header = f"[{msg.timestamp}] {msg.speaker}:"
            else:
                header = f"{msg.speaker}:"

            formatted.append(header)

            # Indent the message
            message_lines = msg.message.split('\n')
            for line in message_lines:
                formatted.append(f"  {line}")

            formatted.append("")  # Empty line between messages

        formatted.append("=" * 60)
        formatted.append(f"Total messages: {len(messages)}")
        formatted.append("=" * 60)

        return '\n'.join(formatted)

    def get_conversation_summary(self, messages: List[ChatMessage]) -> Dict[str, any]:
        """
        Get summary statistics about the conversation.

        Args:
            messages: List of ChatMessage objects

        Returns:
            Dictionary with conversation statistics
        """
        if not messages:
            return {
                'total_messages': 0,
                'participants': [],
                'message_count_by_speaker': {}
            }

        speakers = {}
        for msg in messages:
            speaker = msg.speaker
            if speaker not in speakers:
                speakers[speaker] = 0
            speakers[speaker] += 1

        return {
            'total_messages': len(messages),
            'participants': list(speakers.keys()),
            'message_count_by_speaker': speakers,
            'has_timestamps': any(msg.timestamp for msg in messages)
        }

    def process_text(self, text: str) -> Tuple[bool, str]:
        """
        Process text and format as chat if detected.

        Args:
            text: Input text

        Returns:
            Tuple of (is_chat, formatted_text)
        """
        is_chat = self.is_likely_chat(text)

        if is_chat:
            messages = self.extract_messages(text)
            if messages:
                formatted = self.format_conversation(messages)
                summary = self.get_conversation_summary(messages)

                # Add summary at the top
                summary_text = (
                    f"Detected chat conversation with {summary['total_messages']} messages\n"
                    f"Participants: {', '.join(summary['participants'])}\n\n"
                )

                return True, summary_text + formatted
            else:
                return False, text
        else:
            return False, text
